From 0f6135ae6d97f77c1a737a26f67bf3242940e11c Mon Sep 17 00:00:00 2001
From: Christian Ehrlicher <ch.ehrlicher@gmx.de>
Date: Sat, 13 Sep 2025 16:07:05 +0200
Subject: [PATCH 182/553] QToolButton: always set to MenuButtonPopup when
 action has a menu

The old implementation only set MenuButtonPopup when the corresponding
QAction had a menu *before* it was assigned to a QToolButton but not
afterwards which leads to inconsistencies. Make sure to always set
MenuButtonPopup when the QAction gets a menu. Also mark the todo from
Qt4.2 with a proper Qt7 todo marker so it can be found and fixed during
Qt7 transistion.

Task-number: QTBUG-140148
Change-Id: I4def22e7147049c30a16a7fe88c18591dbb58e88
Reviewed-by: Richard Moe Gustavsen <richard.gustavsen@qt.io>
(cherry picked from commit 6bb5b13b8d74c23135868fc7604bf13f2e159115)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/widgets/widgets/qtoolbutton.cpp | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/src/widgets/widgets/qtoolbutton.cpp b/src/widgets/widgets/qtoolbutton.cpp
index 8957e93f051..e9aea588e66 100644
--- a/src/widgets/widgets/qtoolbutton.cpp
+++ b/src/widgets/widgets/qtoolbutton.cpp
@@ -44,6 +44,7 @@ public:
     void popupTimerDone();
     void updateButtonDown();
     void onMenuTriggered(QAction *);
+    void onDefaultActionChanged();
 #endif
     bool updateHoverControl(const QPoint &pos);
     void onActionTriggered();
@@ -823,6 +824,13 @@ void QToolButtonPrivate::onMenuTriggered(QAction *action)
         emit q->triggered(action);
 }
 
+void QToolButtonPrivate::onDefaultActionChanged()
+{
+    Q_Q(QToolButton);
+    if (defaultAction && defaultAction->menu())
+        q->setPopupMode(QToolButton::MenuButtonPopup);
+}
+
 /*! \enum QToolButton::ToolButtonPopupMode
 
     Describes how a menu should be popped up for tool buttons that has
@@ -914,8 +922,10 @@ void QToolButton::setDefaultAction(QAction *action)
 {
     Q_D(QToolButton);
 #if QT_CONFIG(menu)
-    bool hadMenu = false;
-    hadMenu = d->hasMenu();
+    if (d->defaultAction) {
+        QObjectPrivate::disconnect(d->defaultAction, &QAction::changed, d,
+                                   &QToolButtonPrivate::onDefaultActionChanged);
+    }
 #endif
     d->defaultAction = action;
     if (!action)
@@ -939,12 +949,15 @@ void QToolButton::setDefaultAction(QAction *action)
     setWhatsThis(action->whatsThis());
 #endif
 #if QT_CONFIG(menu)
-    if (action->menu() && !hadMenu) {
+    if (action->menu()) {
+        // ### Qt7 Fixme
         // new 'default' popup mode defined introduced by tool bar. We
         // should have changed QToolButton's default instead. Do that
         // in 4.2.
         setPopupMode(QToolButton::MenuButtonPopup);
     }
+    QObjectPrivate::connect(d->defaultAction, &QAction::changed, d,
+                            &QToolButtonPrivate::onDefaultActionChanged);
 #endif
     setCheckable(action->isCheckable());
     setChecked(action->isChecked());
-- 
2.51.2

